installer/parse_install_yaml.py

```diff
@@ -0,0 +1,258 @@
+
+from lsb_release import get_distro_information
+
+ros_release = environ["TUE_ROS_DISTRO"]
+ubuntu_release = get_distro_information()["CODENAME"]
+
+
+def show_error(error) -> int:
+    print(f"ERROR: {error}")
+    return 1
+
+
+def type_git(install_item: Mapping, allowed_keys: Optional[List[str]] = None) -> str:
+    """
+    Function to check the parsed yaml for install type git and generate the command string.
+
+    The structure of a git install type is:
+    - type: git
+      url: <REPOSITORY_URL>
+      path: [LOCAL_CLONE_PATH]
+      version: [BRANCH_COMMIT_TAG]
+
+    :param install_item: Extracted yaml component corresponding to install type git
+    :param allowed_keys: Additional keys to allow apart from the keys defined in install type git
+    :return: Command string containing repository url and optional arguments target-dir and version
+    """
+
+    assert install_item["type"] == "git", f"Invalid install type '{install_item['type']}' for type 'git'"
+
+    keys = {"type", "url", "path", "version"}
+    if allowed_keys:
+        keys |= set(allowed_keys)
+
+    invalid_keys = [k for k in install_item.keys() if k not in keys]
+
+    if invalid_keys:
+        raise KeyError(f"The following keys are invalid for install type 'git': {invalid_keys}")
+
+    url = install_item.get("url")
+    path = install_item.get("path")
+    version = install_item.get("version")
+
+    if not url:
+        raise KeyError("'url' is a mandatory key for install type 'git'")
+
+    command = url
+
+    if path:
+        command += f" --target-dir={path}"
+
+    if version:
+        command += f" --version={version}"
+
+    return command
+
+
+def catkin_git(source: Mapping) -> str:
+    """
+    Function to generate installation command for catkin git targets from the extracted yaml
+
+    The structure of a catkin git install type is:
+    - type: catkin/ros
+      source:
+        type: git
+        url: <REPOSITORY_URL>
+        path: [LOCAL_CLONE_PATH]
+        version: [BRANCH_COMMIT_TAG]
+        sub-dir: [PACKAGE_SUB_DIRECTORY]
+
+    :param source: Extracted yaml component for the key 'source' corresponding to install type catkin/ros
+    :return: Command string containing arguments to the primary catkin target installation command
+    """
+
+    git_cmd_args = type_git(source, allowed_keys=["sub-dir"])
+    command = f"git {git_cmd_args}"
+
+    sub_dir = source.get("sub-dir")
+    if sub_dir:
+        command += f" --sub-dir={sub_dir}"
+
+    return command
+
+
+def main() -> int:
+    if not 2 <= len(sys.argv) <= 3:
+        return show_error("Usage: parse_install_yaml install.yaml [--now]")
+
+    now = False
+    if len(sys.argv) == 3:
+        if sys.argv[2] == "--now":
+            now = True
+        else:
+            return show_error(f"Unknown option: {sys.argv[2]}")
+
+    try:
+        result = installyaml_parser(sys.argv[1], now)
+    except Exception as e:
+        return show_error(f"{e}")
+
+    if result["commands"]:
```

Comment:
Are we safe for a `KeyError` here?

Reply:
Yes. The default is the return of an empty dictionary.